File: vscode/src/core/localizations/langJson.test.ts
import * as assert from 'assert';
import { parseFlattened, withValue } from './langJson';

const BOM = '﻿';

describe('langJson.parseFlattened', () => {
  it('flattens nested objects to dot notation', () => {
    const map = parseFlattened('{"common":{"ok":"OK","nested":{"deep":"D"}},"buttons":{"login":"Login"}}');
    assert.strictEqual(map.get('common.ok'), 'OK');
    assert.strictEqual(map.get('common.nested.deep'), 'D');
    assert.strictEqual(map.get('buttons.login'), 'Login');
    assert.strictEqual(map.size, 3);
  });

  it('preserves insertion order', () => {
    const map = parseFlattened('{"b":{"z":"1","a":"2"},"a":"3"}');
    assert.deepStrictEqual([...map.keys()], ['b.z', 'b.a', 'a']);
  });

  it('reads raw utf8 characters', () => {
    const map = parseFlattened('{"a":"café","b":"💔"}');
    assert.strictEqual(map.get('a'), 'café');
    assert.strictEqual(map.get('b'), '💔');
  });

  it('decodes json unicode escapes', () => {
    assert.strictEqual(parseFlattened('{"a":"\\u0041\\u00e9"}').get('a'), 'Aé');
  });

  it('keeps escaped quotes verbatim', () => {
    assert.strictEqual(parseFlattened('{"a":"say \\"hi\\""}').get('a'), 'say "hi"');
  });

  it('keeps placeholder tokens verbatim', () => {
    const map = parseFlattened('{"a":"{{xp}} XP","b":"{{terms:terms and conditions}}"}');
    assert.strictEqual(map.get('a'), '{{xp}} XP');
    assert.strictEqual(map.get('b'), '{{terms:terms and conditions}}');
  });

  it('stringifies arrays and maps json null to empty', () => {
    const map = parseFlattened('{"arr":[1,2,"x"],"nul":null,"num":5,"bool":true}');
    assert.strictEqual(map.get('arr'), '[1,2,"x"]');
    assert.strictEqual(map.get('nul'), '');
    assert.strictEqual(map.get('num'), '5');
    assert.strictEqual(map.get('bool'), 'true');
  });

  it('strips a leading BOM', () => {
    assert.strictEqual(parseFlattened(BOM + '{"a":"1"}').get('a'), '1');
  });

  it('duplicate keys keep the last value', () => {
    const map = parseFlattened('{"a":"1","a":"2"}');
    assert.strictEqual(map.get('a'), '2');
    assert.strictEqual(map.size, 1);
  });

  it('throws on malformed json', () => {
    assert.throws(() => parseFlattened('{'));
  });

  it('throws when top level is not an object', () => {
    assert.throws(() => parseFlattened('[1,2,3]'));
  });
});

describe('langJson.withValue', () => {
  it('updates an existing key', () => {
    const updated = withValue('{"common":{"ok":"OK"}}', 'common.ok', 'Okay');
    assert.strictEqual(parseFlattened(updated).get('common.ok'), 'Okay');
  });

  it('inserts a missing nested key', () => {
    const updated = withValue('{"common":{"ok":"OK"}}', 'common.deep.new', 'Hi');
    const map = parseFlattened(updated);
    assert.strictEqual(map.get('common.deep.new'), 'Hi');
    assert.strictEqual(map.get('common.ok'), 'OK');
  });

  it('preserves a single trailing newline', () => {
    const updated = withValue('{}\n', 'a', '1');
    assert.ok(updated.endsWith('\n'));
    assert.ok(!updated.replace(/\n$/, '').endsWith('\n'));
  });

  it('omits a trailing newline when the source has none', () => {
    assert.ok(!withValue('{"a":"x"}', 'a', 'y').endsWith('\n'));
  });

  it('keeps placeholders and emoji unescaped', () => {
    assert.ok(withValue('{"a":"x"}', 'b', '{{xp}} 💔').includes('{{xp}} 💔'));
  });

  it('writes apostrophes and html chars literally', () => {
    const apos = withValue('{"a":"x"}', 'score', "Score're");
    assert.ok(apos.includes("Score're"));
    assert.ok(!apos.includes('\\u0027')); // literal backslash-u-0027 must be absent

    const html = withValue('{"a":"x"}', 'b', 'a < b & c > d = e');
    assert.ok(html.includes('a < b & c > d = e'));
  });

  it('updates a flat dotted key in place instead of nesting a duplicate', () => {
    const updated = withValue('{"login.email":"E-mail"}', 'login.email', 'Email');
    const map = parseFlattened(updated);
    assert.strictEqual(map.get('login.email'), 'Email');
    assert.strictEqual(map.size, 1);
    assert.ok(updated.includes('"login.email"')); // still stored flat, not exploded to nested
  });

  it('updates through an object name that itself contains a dot', () => {
    const updated = withValue('{"a.b":{"c":"old"}}', 'a.b.c', 'new');
    const map = parseFlattened(updated);
    assert.strictEqual(map.get('a.b.c'), 'new');
    assert.strictEqual(map.size, 1);
  });

  it('never replaces a scalar intermediate - the new key is written flat', () => {
    const updated = withValue('{"login":"Anmelden"}', 'login.email', 'E-Mail');
    const map = parseFlattened(updated);
    assert.strictEqual(map.get('login'), 'Anmelden'); // the existing translation survives
    assert.strictEqual(map.get('login.email'), 'E-Mail');
  });

  it('refuses to overwrite a nested object with a string', () => {
    assert.throws(() => withValue('{"login":{"email":"x"}}', 'login', 'y'));
  });
});

File: intellij/src/test/kotlin/dev/nylo/plugin/localizations/json/LangJsonTest.kt
package dev.nylo.plugin.localizations.json

import com.google.gson.JsonParseException
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class LangJsonTest {

    @Test
    fun `flattens nested objects to dot notation`() {
        val map = LangJson.parseFlattened(
            """{"common":{"ok":"OK","nested":{"deep":"D"}},"buttons":{"login":"Login"}}""",
        )
        assertEquals("OK", map["common.ok"])
        assertEquals("D", map["common.nested.deep"])
        assertEquals("Login", map["buttons.login"])
        assertEquals(3, map.size)
    }

    @Test
    fun `preserves insertion order`() {
        val map = LangJson.parseFlattened("""{"b":{"z":"1","a":"2"},"a":"3"}""")
        assertEquals(listOf("b.z", "b.a", "a"), map.keys.toList())
    }

    @Test
    fun `reads raw utf8 characters`() {
        val map = LangJson.parseFlattened("""{"a":"café","b":"💔"}""")
        assertEquals("café", map["a"])
        assertEquals("💔", map["b"])
    }

    @Test
    fun `decodes json unicode escapes`() {
        val bs = 0x5C.toChar() // backslash, kept out of source as a literal escape
        val json = "{\"a\":\"" + bs + "u0041" + bs + "u00e9\"}" // {"a":"Aé"}
        assertEquals("Aé", LangJson.parseFlattened(json)["a"])
    }

    @Test
    fun `keeps escaped quotes verbatim`() {
        // {"a":"say \"hi\""}
        val map = LangJson.parseFlattened("{\"a\":\"say \\\"hi\\\"\"}")
        assertEquals("say \"hi\"", map["a"])
    }

    @Test
    fun `keeps placeholder tokens verbatim`() {
        val map = LangJson.parseFlattened(
            """{"a":"{{xp}} XP","b":"{{terms:terms and conditions}}"}""",
        )
        assertEquals("{{xp}} XP", map["a"])
        assertEquals("{{terms:terms and conditions}}", map["b"])
    }

    @Test
    fun `stringifies arrays and maps json null to empty`() {
        val map = LangJson.parseFlattened("""{"arr":[1,2,"x"],"nul":null,"num":5,"bool":true}""")
        assertEquals("[1,2,\"x\"]", map["arr"])
        assertEquals("", map["nul"])
        assertEquals("5", map["num"])
        assertEquals("true", map["bool"])
    }

    @Test
    fun `strips a leading BOM`() {
        val bom = 0xFEFF.toChar()
        val map = LangJson.parseFlattened(bom + """{"a":"1"}""")
        assertEquals("1", map["a"])
    }

    @Test
    fun `duplicate keys keep the last value`() {
        val map = LangJson.parseFlattened("""{"a":"1","a":"2"}""")
        assertEquals("2", map["a"])
        assertEquals(1, map.size)
    }

    @Test(expected = JsonParseException::class)
    fun `throws on malformed json`() {
        LangJson.parseFlattened("{")
    }

    @Test(expected = IllegalArgumentException::class)
    fun `throws when top level is not an object`() {
        LangJson.parseFlattened("""[1,2,3]""")
    }

    @Test
    fun `withValue updates an existing key`() {
        val updated = LangJson.withValue("""{"common":{"ok":"OK"}}""", "common.ok", "Okay")
        assertEquals("Okay", LangJson.parseFlattened(updated)["common.ok"])
    }

    @Test
    fun `withValue inserts a missing nested key`() {
        val updated = LangJson.withValue("""{"common":{"ok":"OK"}}""", "common.deep.new", "Hi")
        val map = LangJson.parseFlattened(updated)
        assertEquals("Hi", map["common.deep.new"])
        assertEquals("OK", map["common.ok"])
    }

    @Test
    fun `withValue preserves a single trailing newline`() {
        val updated = LangJson.withValue("{}\n", "a", "1")
        assertTrue(updated.endsWith("\n"))
        assertFalse(updated.removeSuffix("\n").endsWith("\n"))
    }

    @Test
    fun `withValue omits a trailing newline when the source has none`() {
        assertFalse(LangJson.withValue("""{"a":"x"}""", "a", "y").endsWith("\n"))
    }

    @Test
    fun `withValue keeps placeholders and emoji unescaped`() {
        val updated = LangJson.withValue("""{"a":"x"}""", "b", "{{xp}} 💔")
        assertTrue(updated.contains("{{xp}} 💔"))
    }

    @Test
    fun `withValue writes apostrophes and html chars literally`() {
        val apos = LangJson.withValue("""{"a":"x"}""", "score", "Score're")
        assertTrue(apos.contains("Score're"))
        assertFalse(apos.contains("\\u0027"))   // literal backslash-u-0027 must be absent

        val html = LangJson.withValue("""{"a":"x"}""", "b", "a < b & c > d = e")
        assertTrue(html.contains("a < b & c > d = e"))
    }

    @Test
    fun `withValue updates a flat dotted key in place instead of nesting a duplicate`() {
        val updated = LangJson.withValue("""{"login.email":"E-mail"}""", "login.email", "Email")
        val map = LangJson.parseFlattened(updated)
        assertEquals("Email", map["login.email"])
        assertEquals(1, map.size)
        assertTrue(updated.contains("\"login.email\"")) // still stored flat, not exploded to nested
    }

    @Test
    fun `withValue updates through an object name that itself contains a dot`() {
        val updated = LangJson.withValue("""{"a.b":{"c":"old"}}""", "a.b.c", "new")
        val map = LangJson.parseFlattened(updated)
        assertEquals("new", map["a.b.c"])
        assertEquals(1, map.size)
    }

    @Test
    fun `withValue never replaces a scalar intermediate - the new key is written flat`() {
        val updated = LangJson.withValue("""{"login":"Anmelden"}""", "login.email", "E-Mail")
        val map = LangJson.parseFlattened(updated)
        assertEquals("Anmelden", map["login"]) // the existing translation survives
        assertEquals("E-Mail", map["login.email"])
    }

    @Test(expected = IllegalArgumentException::class)
    fun `withValue refuses to overwrite a nested object with a string`() {
        LangJson.withValue("""{"login":{"email":"x"}}""", "login", "y")
    }
}
